### ISSUE TITLE
Multi machine and multi disk per machine fails

### ISSUE THREAD

**Issue by javiroman** [vagrant-libvirt/vagrant-libvirt#651]
### Steps to reproduce
1. Using this Vagrant file:
```
https://github.com/bigcontainer/bigcont/blob/master/testbed/ocptestbed/Vagrantfile
```
2. VAGRANT_NO_PARALLEL=true vagrant up --no-provision

3. The affected part of the Vagrant file:

```
config.vm.define "#{vm_name}" do |node|                                                                                                                                                   
    │ node.vm.hostname = "#{vm_name}"                                                                                                                                                         
    │ #node.vm.network :private_network, ip: "10.10.10.10#{i}", auto_config: false                                                                                                            
    │ config.vm.provider :libvirt do |domain|                                                                                                                                                 
    │   domain.uri = 'qemu+unix:///system'                                                                                                                                                    
    │   domain.driver = 'kvm'                                                                                                                                                                 
    │   domain.memory = "#{vm_memory}".to_i                                                                                                                                                   
    │   domain.cpus = 1                                                                                                                                                                       
    │   # workers nodes simulate JBOD disk arrays (infra node excluded)                                                                                                                       
    │   domain.storage 0                                                                                                                                                                      
    │   if "#{node.vm.hostname}" == "node1"                                                                                                                                                   
    │   │   domain.storage :file, :size => '10G'                                                                                                                                              
    │   │   domain.storage :file, :size => '10G'                                                                                                                                              
    │   │   domain.storage :file, :size => '10G'                                                                                                                                              
    │   elsif"#{node.vm.hostname}" == "node2"                                                                                                                                                 
    │   │   domain.storage :file, :size => '10G'                                                                                                                                              
    │   │   domain.storage :file, :size => '10G'                                                                                                                                              
    │   elsif"#{node.vm.hostname}" == "node3"                                                                                                                                                 
    │   │   domain.storage :file, :size => '10G'                                                                                                                                              
    │   end                                                                                                                                                                                   
    │ end
```

### Expected behaviour

Create 3 disks for "node1", 2 disks for "node2" and one additional disk for "node3"

### Actual behaviour

```
==> node1:  -- Disks:         vdb(qcow2,10G), vdc(qcow2,10G), vdd(qcow2,10G)                                                                                                                  
==> node1:  -- Disk(vdb):     /var/lib/libvirt/images/ocptestbed_node1-vdb.qcow2                                                                                                              
==> node1:  -- Disk(vdc):     /var/lib/libvirt/images/ocptestbed_node1-vdc.qcow2                                                                                                              
==> node1:  -- Disk(vdd):     /var/lib/libvirt/images/ocptestbed_node1-vdd.qcow2 `
```

And the wrong disk creation in the "node2"

```
==> node2:  -- Disks:         vdb(qcow2,10G), vdc(qcow2,10G), vdd(qcow2,10G), vdb(qcow2,10G), vdc(qcow2,10G)                                                                                  
==> node2:  -- Disk(vdb):     /var/lib/libvirt/images/ocptestbed_node2-vdb.qcow2                                                                                                              
==> node2:  -- Disk(vdc):     /var/lib/libvirt/images/ocptestbed_node2-vdc.qcow2                                                                                                              
==> node2:  -- Disk(vdd):     /var/lib/libvirt/images/ocptestbed_node2-vdd.qcow2                                                                                                              
==> node2:  -- Disk(vdb):     /var/lib/libvirt/images/ocptestbed_node2-vdb.qcow2 Not created - using existed.                                                                                 
==> node2:  -- Disk(vdc):     /var/lib/libvirt/images/ocptestbed_node2-vdc.qcow2 Not created - using existed. 
```

### System configuration

Fedora 24 with

```
$ vagrant --version
Vagrant 1.8.1
$ vagrant plugin list
vagrant-libvirt (0.0.35)
```

**Output of `VAGRANT_LOG=debug vagrant ... --provider=libvirt`**
```
Bringing machine 'mn' up with 'libvirt' provider...                                                                                                                                           
Bringing machine 'master' up with 'libvirt' provider...                                                                                                                                       
Bringing machine 'node1' up with 'libvirt' provider...                                                                                                                                        
Bringing machine 'node2' up with 'libvirt' provider...                                                                                                                                        
Bringing machine 'node3' up with 'libvirt' provider...                                                                                                                                        
Bringing machine 'node4' up with 'libvirt' provider...                                                                                                                                        
==> mn: Creating image (snapshot of base box volume).                                                                                                                                         
==> mn: Creating domain with the following settings...                                                                                                                                        
==> mn:  -- Name:              ocptestbed_mn                                                                                                                                                  
==> mn:  -- Domain type:       kvm                                                                                                                                                            
==> mn:  -- Cpus:              1                                                                                                                                                              
==> mn:  -- Memory:            1024M                                                                                                                                                          
==> mn:  -- Management MAC:                                                                                                                                                                   
==> mn:  -- Loader:                                                                                                                                                                           
==> mn:  -- Base box:          centos/7                                                                                                                                                       
==> mn:  -- Storage pool:      default                                                                                                                                                        
==> mn:  -- Image:             /var/lib/libvirt/images/ocptestbed_mn.img (41G)                                                                                                                
==> mn:  -- Volume Cache:      default                                                                                                                                                        
==> mn:  -- Kernel:                                                                                                                                                                           
==> mn:  -- Initrd:                                                                                                                                                                           
==> mn:  -- Graphics Type:     vnc                                                                                                                                                            
==> mn:  -- Graphics Port:     5900                                                                                                                                                           
==> mn:  -- Graphics IP:       127.0.0.1                                                                                                                                                      
==> mn:  -- Graphics Password: Not defined                                                                                                                                                    
==> mn:  -- Video Type:        cirrus                                                                                                                                                         
==> mn:  -- Video VRAM:        9216                                                                                                                                                           
==> mn:  -- Keymap:            en-us                                                                                                                                                          
==> mn:  -- TPM Path:                                                                                                                                                                         
==> mn:  -- INPUT:             type=mouse, bus=ps2                                                                                                                                            
==> mn:  -- Command line :                                                                                                                                                                    
==> mn: Creating shared folders metadata...                                                                                                                                                   
==> mn: Starting domain.                                                                                                                                                                      
==> mn: Waiting for domain to get an IP address...                                                                                                                                            
==> mn: Waiting for SSH to become available...                                                                                                                                                
    mn:                                                                                                                                                                                       
    mn: Vagrant insecure key detected. Vagrant will automatically replace                                                                                                                     
    mn: this with a newly generated keypair for better security.                                                                                                                              
    mn:                                                                                                                                                                                       
    mn: Inserting generated public key within guest...                                                                                                                                        
    mn: Removing insecure key from the guest if it's present...                                                                                                                               
    mn: Key inserted! Disconnecting and reconnecting using new SSH key...                                                                                                                     
==> mn: Setting hostname... 
==> mn: Configuring and enabling network interfaces...                                                                                                                                        
==> mn: Machine not provisioned because `--no-provision` is specified.                                                                                                                        
==> master: Creating image (snapshot of base box volume).                                                                                                                                     
==> master: Creating domain with the following settings...                                                                                                                                    ==> master:  -- Name:              ocptestbed_master                                                                                                                                          
==> master:  -- Domain type:       kvm                                                                                                                                                        
==> master:  -- Cpus:              1                                                                                                                                                          
==> master:  -- Memory:            2048M                                                                                                                                                      
==> master:  -- Management MAC:                                                                                                                                                               
==> master:  -- Loader:                                                                                                                                                                       
==> master:  -- Base box:          centos/7                                                                                                                                                   
==> master:  -- Storage pool:      default                                                                                                                                                    
==> master:  -- Image:             /var/lib/libvirt/images/ocptestbed_master.img (41G)                                                                                                        
==> master:  -- Volume Cache:      default                                                                                                                                                    
==> master:  -- Kernel:                                                                                                                                                                       
==> master:  -- Initrd:                                                                                                                                                                       
==> master:  -- Graphics Type:     vnc                                                                                                                                                        
==> master:  -- Graphics Port:     5900                                                                                                                                                       
==> master:  -- Graphics IP:       127.0.0.1                                                                                                                                                  
==> master:  -- Graphics Password: Not defined                                                                                                                                                
==> master:  -- Video Type:        cirrus                                                                                                                                                     
==> master:  -- Video VRAM:        9216                                                                                                                                                       
==> master:  -- Keymap:            en-us                                                                                                                                                      
==> master:  -- TPM Path:                                                                                                                                                                     
==> master:  -- INPUT:             type=mouse, bus=ps2                                                                                                                                        
==> master:  -- Command line :                                                                                                                                                                
==> master: Creating shared folders metadata...                                                                                                                                               
==> master: Starting domain.                                                                                                                                                                  
==> master: Waiting for domain to get an IP address...                                                                                                                                        
==> master: Waiting for SSH to become available...                                                                                                                                            
    master:                                                                                                                                                                                   
    master: Vagrant insecure key detected. Vagrant will automatically replace                                                                                                                 
    master: this with a newly generated keypair for better security.                                                                                                                          
    master:                                                                                                                                                                                   
    master: Inserting generated public key within guest...                                                                                                                                    
    master: Removing insecure key from the guest if it's present...                                                                                                                           
    master: Key inserted! Disconnecting and reconnecting using new SSH key...                                                                                                                 
==> master: Setting hostname... 
==> master: Configuring and enabling network interfaces...                                                                                                                                    
==> master: Machine not provisioned because `--no-provision` is specified.                                                                                                                    
==> node1: Creating image (snapshot of base box volume).                                                                                                                                      
==> node1: Creating domain with the following settings...                                                                                                                                     
==> node1:  -- Name:              ocptestbed_node1                                                                                                                                            
==> node1:  -- Domain type:       kvm                                                                                                                                                         
==> node1:  -- Cpus:              1                                                                                                                                                           
==> node1:  -- Memory:            2048M                                                                                                                                                       
==> node1:  -- Management MAC:                                                                                                                                                                
==> node1:  -- Loader:                                                                                                                                                                        
==> node1:  -- Base box:          centos/7                                                                                                                                                    
==> node1:  -- Storage pool:      default                                                                                                                                                     
==> node1:  -- Image:             /var/lib/libvirt/images/ocptestbed_node1.img (41G)                                                                                                          
==> node1:  -- Volume Cache:      default                                                                                                                                                     
==> node1:  -- Kernel:                                                                                                                                                                        
==> node1:  -- Initrd:                                                                                                                                                                        
==> node1:  -- Graphics Type:     vnc                                                                                                                                                         
==> node1:  -- Graphics Port:     5900                                                                                                                                                        
==> node1:  -- Graphics IP:       127.0.0.1                                                                                                                                                   
==> node1:  -- Graphics Password: Not defined                                                                                                                                                 
==> node1:  -- Video Type:        cirrus                                                                                                                                                      
==> node1:  -- Video VRAM:        9216                                                                                                                                                        
==> node1:  -- Keymap:            en-us                                                                                                                                                       
==> node1:  -- TPM Path:                                                                                                                                                                      
==> node1:  -- Disks:         vdb(qcow2,10G), vdc(qcow2,10G), vdd(qcow2,10G)                                                                                                                  
==> node1:  -- Disk(vdb):     /var/lib/libvirt/images/ocptestbed_node1-vdb.qcow2                                                                                                              
==> node1:  -- Disk(vdc):     /var/lib/libvirt/images/ocptestbed_node1-vdc.qcow2                                                                                                              
==> node1:  -- Disk(vdd):     /var/lib/libvirt/images/ocptestbed_node1-vdd.qcow2                                                                                                              
==> node1:  -- INPUT:             type=mouse, bus=ps2                                                                                                                                         
==> node1:  -- Command line :                                                                                                                                                                 
==> node1: Creating shared folders metadata...                                                                                                                                                
==> node1: Starting domain.                                                                                                                                                                   
==> node1: Waiting for domain to get an IP address...                                                                                                                                         
==> node1: Waiting for SSH to become available...                                                                                                                                             
    node1:                                                                                                                                                                                    
    node1: Vagrant insecure key detected. Vagrant will automatically replace                                                                                                                  
    node1: this with a newly generated keypair for better security.                                                                                                                           
    node1:                                                                                                                                                                                    
    node1: Inserting generated public key within guest...                                                                                                                                     
    node1: Removing insecure key from the guest if it's present...                                                                                                                            
    node1: Key inserted! Disconnecting and reconnecting using new SSH key...                                                                                                                  
==> node1: Setting hostname... 
==> node1: Configuring and enabling network interfaces...                                                                                                                                     
==> node1: Machine not provisioned because `--no-provision` is specified.                                                                                                                     
==> node2: Creating image (snapshot of base box volume).                                                                                                                                      
==> node2: Creating domain with the following settings...                                                                                                                                     
==> node2:  -- Name:              ocptestbed_node2                                                                                                                                            
==> node2:  -- Domain type:       kvm                                                                                                                                                         
==> node2:  -- Cpus:              1                                                                                                                                                           
==> node2:  -- Memory:            2048M                                                                                                                                                       
==> node2:  -- Management MAC:                                                                                                                                                                
==> node2:  -- Loader:                                                                                                                                                                        
==> node2:  -- Base box:          centos/7                                                                                                                                                    
==> node2:  -- Storage pool:      default                                                                                                                                                     
==> node2:  -- Image:             /var/lib/libvirt/images/ocptestbed_node2.img (41G)                                                                                                          
==> node2:  -- Volume Cache:      default                                                                                                                                                     
==> node2:  -- Kernel:                                                                                                                                                                        
==> node2:  -- Initrd:                                                                                                                                                                        
==> node2:  -- Graphics Type:     vnc                                                                                                                                                         
==> node2:  -- Graphics Port:     5900                                                                                                                                                        
==> node2:  -- Graphics IP:       127.0.0.1                                                                                                                                                   
==> node2:  -- Graphics Password: Not defined                                                                                                                                                 
==> node2:  -- Video Type:        cirrus                                                                                                                                                      
==> node2:  -- Video VRAM:        9216                                                                                                                                                        
==> node2:  -- Keymap:            en-us                                                                                                                                                       
==> node2:  -- TPM Path:                                                                                                                                                                      
==> node2:  -- Disks:         vdb(qcow2,10G), vdc(qcow2,10G), vdd(qcow2,10G), vdb(qcow2,10G), vdc(qcow2,10G)                                                                                  
==> node2:  -- Disk(vdb):     /var/lib/libvirt/images/ocptestbed_node2-vdb.qcow2                                                                                                              
==> node2:  -- Disk(vdc):     /var/lib/libvirt/images/ocptestbed_node2-vdc.qcow2                                                                                                              
==> node2:  -- Disk(vdd):     /var/lib/libvirt/images/ocptestbed_node2-vdd.qcow2                                                                                                              
==> node2:  -- Disk(vdb):     /var/lib/libvirt/images/ocptestbed_node2-vdb.qcow2 Not created - using existed.                                                                                 
==> node2:  -- Disk(vdc):     /var/lib/libvirt/images/ocptestbed_node2-vdc.qcow2 Not created - using existed.                                                                                 
==> node2:  -- INPUT:             type=mouse, bus=ps2                                                                                                                                         
==> node2:  -- Command line :                                                                                                                                                                 
==> node2: Creating shared folders metadata...                                                                                                                                                
==> node2: Starting domain.                                                                                                                                                                   
There was an error talking to Libvirt. The error message is shown                                                                                                                             
below:                                                                                                                                                                                        
                                                                                                                                                                                              
Call to virDomainCreateWithFlags failed: XML error: target 'vdb' duplicated for disk sources '/var/lib/libvirt/images/ocptestbed_node2-vdb.qcow2' and '/var/lib/libvirt/images/ocptestbed_node2-vdb.qcow2' 
```

**Comment by javiroman** [comment]
Probably the "disks" array must be initialized each VM processed. In this wrong behavior the "disks" array is doing any kind of "append" of the disks.

**Comment by javiroman** [comment]
The only workaround is run the VM individually:

vagrant up mn master node1
vagrant up node1
vagrant up node2
vagrant up node3

**Comment by infernix** [comment]
Can reproduce behaviour but not sure yet where this is going wrong.

**Comment by electrofelix** [comment]
You are redefining the storage for config.vm.provider, which is global, when processing each node definition instead of the provider config for that specific vm definition.

**Comment by electrofelix** [comment]
There is a big in how storage configuration is merged across configs in how the device name is determined which is resulting in some additional issues here as well, thus confusing the issue even more, but your conduit is incorrect anyway

**Comment by infernix** [comment]
Can confirm, this works, even in parallel:

```diff
diff --git a/Vagrantfile.orig b/Vagrantfile
index 3466b34..9d2392e 100644
--- a/Vagrantfile.orig
+++ b/Vagrantfile
@@ -47,7 +47,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
     config.vm.define "#{vm_name}" do |node|
       node.vm.hostname = "#{vm_name}"
       #node.vm.network :private_network, ip: "10.10.10.10#{i}", auto_config: false 
-      config.vm.provider :libvirt do |domain|
+      node.vm.provider :libvirt do |domain|
         domain.uri = 'qemu+unix:///system'
         domain.driver = 'kvm'
         domain.memory = "#{vm_memory}".to_i
```

Therefore closing this. @electrofelix I'm not seeing the other issue you mentioned; let's make a new issue if it's worth pursuing? :+1:

**Comment by electrofelix** [comment]
@infernix will do, just for clarity here what should of happened with updating the global storage was that with:

NUM_HOSTS=4
vms = mn, master, node1 & node2
result all 4 would have gotten 4 additional disks assigned

NUM_HOSTS=6
vms = mn, master, node1, node2, node3, & node4
result all 4 would have gotten 10 additional disks assigned

It probably would have been a little clearer to the user that they were changing a global setting rather than per node setting in that case. Essentially the internal disk arrays for enumerating storage devices added are not merged across configuration blocks correctly in vagrant-libvirt, partly due to how the device is assigned are population time rather than finalize resulting in any merge of arrays potentially having duplicate disk identifiers which will overwrite one another.
